core/silkworm/common/object_pool.hpp

```diff
@@ -14,51 +14,77 @@
    limitations under the License.
 */
 
-#ifndef SILKWORM_COMMON_OBJECTPOOL_HPP_
-#define SILKWORM_COMMON_OBJECTPOOL_HPP_
+#ifndef SILKWORM_COMMON_OBJECT_POOL_HPP_
+#define SILKWORM_COMMON_OBJECT_POOL_HPP_
 
 #include <memory>
 #include <stack>
+#include <vector>
+
+#ifndef __wasm__
+#include <mutex>
+#endif
+
+#include <gsl/pointers>
+
+#ifndef __wasm__
+#define SILKWORM_DETAIL_OBJECT_POOL_GUARD \
+    std::unique_lock<std::mutex> lock;    \
+    if (thread_safe_) {                   \
+        lock = std::unique_lock{mutex_};  \
+    }
+#else
+#define SILKWORM_DETAIL_OBJECT_POOL_GUARD
+#endif
 
 namespace silkworm {
 
 template <class T, class TDtor = std::default_delete<T>>
 class ObjectPool {
   public:
-    using ptr_t = std::unique_ptr<T, TDtor>;
-
-    ObjectPool() = default;
+    explicit ObjectPool(bool thread_safe = false) : thread_safe_{thread_safe} {}
 
-    virtual ~ObjectPool() = default;
+    // Not copyable nor movable
+    ObjectPool(const ObjectPool&) = delete;
+    ObjectPool& operator=(const ObjectPool&) = delete;
 
-    void add(T*& t) {
+    void add(gsl::owner<T*> t) {
+        SILKWORM_DETAIL_OBJECT_POOL_GUARD
         pool_.push({t, TDtor()});
-        t = nullptr;
     }
 
-    T* acquire() {
+    gsl::owner<T*> acquire() {
+        SILKWORM_DETAIL_OBJECT_POOL_GUARD
         if (pool_.empty()) {
             return nullptr;
         }
-        T* ret(pool_.top().release());
+        gsl::owner<T*> ret(pool_.top().release());
```

Comment:
`std::unique_ptr` is probably a good fit for the use of `ObjectPool` for evm states,  but it might be a bad fit for its use in `node/silkworm/db/mdbx.cpp`. Prior to PR #605 `ObjectPool` was actually using `std::unique_ptr`. In this PR I'm simply using `gsl::owner` for formalization of the changes made in PR #605.

Reply:
And core doesn't use exceptions, so no need to worry about them.